packages/devtools_app/lib/src/memory/memory_treemap.dart

```diff
@@ -0,0 +1,266 @@
+// Copyright 2020 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart' hide TextStyle;
+import 'package:flutter/widgets.dart' hide TextStyle;
+import 'package:provider/provider.dart';
+
+import '../auto_dispose_mixin.dart';
+import '../charts/treemap.dart';
+
+import 'memory_controller.dart';
+import 'memory_graph_model.dart';
+import 'memory_utils.dart';
+
+class MemoryTreemap extends StatefulWidget {
+  const MemoryTreemap(this.controller);
+
+  final MemoryController controller;
+
+  @override
+  MemoryTreemapState createState() => MemoryTreemapState(controller);
+}
+
+class MemoryTreemapState extends State<MemoryTreemap> with AutoDisposeMixin {
+  MemoryTreemapState(this.controller);
+
+  InstructionsSize sizes;
+
+  Map<String, Function> callbacks = {};
+
+  MemoryController controller;
+
+  Widget snapshotDisplay;
+
+  TreemapNode root;
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+
+    // TODO(terry): Unable to short-circuit need to investigate why?
+    controller = Provider.of<MemoryController>(context);
+
+    sizes = InstructionsSize.fromSnapshot(controller);
+
+    root = sizes.root;
+
+    cancel();
+
+    addAutoDisposeListener(controller.selectedSnapshotNotifier, () {
+      setState(() {
+        controller.computeAllLibraries(true, true);
+
+        sizes = InstructionsSize.fromSnapshot(controller);
+      });
+    });
+
+    addAutoDisposeListener(controller.filterNotifier, () {
+      setState(() {
+        controller.computeAllLibraries(true, true);
+      });
+    });
+    // TODO(peterdjlee): Need to check if applicable to treemap.
+    // addAutoDisposeListener(controller.selectTheSearchNotifier, () {
+    //   setState(() {
+    //     if (_trySelectItem()) {
+    //       closeAutoCompleteOverlay();
+    //     }
+    //   });
+    // });
+
+    // addAutoDisposeListener(controller.searchNotifier, () {
+    //   setState(() {
+    //     if (_trySelectItem()) {
+    //       closeAutoCompleteOverlay();
+    //     }
+    //   });
+    // });
+
+    addAutoDisposeListener(controller.searchAutoCompleteNotifier, () {
+      setState(autoCompleteOverlaySetState(controller, context));
+    });
+  }
+
+  void _onRootChanged(TreemapNode newRoot) {
+    setState(() {
+      root = newRoot;
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    if (sizes != null) {
+      return LayoutBuilder(
+        builder: (context, constraints) {
+          return Treemap.fromRoot(
+            rootNode: root,
+            levelsVisible: 2,
+            outermostLevel: 2,
```

Comment:
@jacob314 I feel like it might not be the best design to make a developer change both `levelsVisible` and `outermostLevel` since both values need to be the same for the code to work as expected. What would be a better solution?

Reply:
here's a cleaner design:
specify `levelsVisible` and `isOutermostLevel` as it looks like you are only using outermostLevel to check whether you are currently at the `outermostLevel`.